[MBL-1899] Checkout screen PLOT Badge

jovaniks opened the pull request:
<!-- This template is **just a guide**, delete any and all parts which you don't need! -->

# 📲 What

This PR adds a **Pledge Over Time (PLOT) badge** under the "Your pledge" section when PLOT is selected in the collection plan. It also updates the payment label and status label copy for PLOT-enabled pledges.

# 🤔 Why

Some background context on why the change is needed.

# 🛠 How

- Added a **Pledge Over Time badge** to be displayed under the "Your pledge" section.
- Updated the **payments label** to show "charged as 4 payments."
- Updated the **status label copy** to display dynamic text:

   > "If the project reaches its funding goal, the first charge of $20 will be collected on March 15, 2024."

# 👀 See

- [MBL-1899](https://kickstarter.atlassian.net/browse/MBL-1899)

![Simulator Screen Recording - iPhone SE (3rd generation) - 2024-12-16 at 10 19 42](https://github.com/user-attachments/assets/927a0fe1-b17e-46a5-9e5c-de0a69ef2496)

# ✅ Acceptance criteria

- [ ] The **Pledge Over Time badge** is displayed under "Your pledge" when PLOT is selected.
- [ ] The payments label reads: "charged as 4 payments."
- [ ] The status label copy displays the correct amount and scheduled collection date
    - Example: "If the project reaches its funding goal, the first charge of $20 will be collected on March 15, 2024."

# ⏰ TODO

- [ ] Using real data from API when availble ([MBL-1838](https://kickstarter.atlassian.net/browse/MBL-1838))

[MBL-1899]: https://kickstarter.atlassian.net/browse/MBL-1899?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[MBL-1838]: https://kickstarter.atlassian.net/browse/MBL-1838?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ifosli commented on Kickstarter-iOS/Features/PledgePaymentMethods/Controller/PostCampaignPledgeRewardsSummaryTotalViewController.swift:
please add a snapshot test here. It looks like this class doesn't have snapshot test coverage yet (and neither do the main view controllers), so it'd be nice to add both a test where PLOT ui shows and one where it doesn't. We really can't keep pushing off adding test coverage.

ifosli commented on Library/ViewModels/NoShippingPledgeViewModel.swift:
```diff
@@ -987,23 +964,51 @@ public class NoShippingPledgeViewModel: NoShippingPledgeViewModelType, NoShippin
+      additionalPledgeAmount,
+      shippingSummaryViewData,
+      rewards,
+      self.pledgeOverTimeConfigData
```
it looks like this signal will only emit if the PLOT ui is visible. (If one signal from `combineLatest` never emits, the input will never all exist and the signal won't do anything.) You'll want to make sure this emits `nil` when the UI doesn't show.

jovaniks replied:
Confirmed! I ensured that the signal emits even when the input is nil, and I verified it works as expected. Let me know if there’s anything else you’d like me to double-check.

ifosli commented on Library/ViewModels/NoShippingPostCampaignCheckoutViewModel.swift:
```diff
@@ -238,7 +238,8 @@ public class NoShippingPostCampaignCheckoutViewModel: NoShippingPostCampaignChec
         confirmationLabelHidden: true,
-        pledgeHasNoReward: pledgeHasNoRewards(rewards: rewardsData.rewards)
+        pledgeHasNoReward: pledgeHasNoRewards(rewards: rewardsData.rewards),
+        pledgeOverTimeData: nil
```
Is pledge over time an option for late pledges? If it is, why is this hardcoded to nil? If it's not, why are the `PostCampaignPledgeRewardsSummaryTotalViewController` changes required?

jovaniks replied:
In this version (v1), we won’t support late pledges, but we will in V2. For now, it’s hardcoded to `nil` because we don’t need it here at the moment.

Regarding your last question, the changes in `PostCampaignPledgeRewardsSummaryTotalViewController` are required because this controller is used on the checkout screen to display the `Pledge Amount`, just below the `Your Pledge` table.

<img width="320" alt="Screenshot 2024-12-13 at 11 36 17 a m" src="https://github.com/user-attachments/assets/bdb6f7db-bde4-47a5-8890-e5d325167230" />

![image](https://github.com/user-attachments/assets/5f11d6c2-630e-4d01-ad17-6c9a912ab9b3)

ifosli reviewed (changes requested):
Overall, good start! Left a couple comments where it looks like behavior might be broken when the flag is off (please check!) and what we're trying to do with late pledges.

nativeksr commented:
<!--
  0 Errors
  1 Warning: Big PR
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

jovaniks commented:
- Refactored the way `PostCampaignPledgeRewardsSummaryTotalViewController` is notified when `PledgeOverTime` is selected, ensuring the Badge view is displayed properly.
- Added unit tests for `PostCampaignPledgeRewardsSummaryTotalViewController`.
- Added test cases for `NoShippingPledgeViewController`.

This refactor improves the logic flow for `PledgeOverTime` handling and adds coverage for key components.

scottkicks commented on Library/ViewModels/PledgeSummaryViewModel.swift:
```diff
@@ -103,13 +113,24 @@ public class PledgeSummaryViewModel: PledgeSummaryViewModelType,
+
+    // TODO: add strings translations [MBL-1860](https://kickstarter.atlassian.net/browse/MBL-1860)
+    self.pledgeOverTimeChargesText = pledgeOverTimeData.skipNil()
+      .map { "charged as \($0.paymentIncrements.count) payments" }
```
I think the latest strings are available on `main` so we can update this now.

jovaniks replied:
Ohh, great catch! Thanks for pointing that out. My question now is: should we also update the test cases to use the strings?

scottkicks commented on Kickstarter-iOS/Features/PledgePaymentMethods/Controller/PostCampaignPledgeRewardsSummaryTotalViewController.swift:
If it isn't too much work it'd be nice to update this file name since it's being used in both the Crowdfund and Late Pledge (Previously Post Campaign Pledge) flows. Maybe `PledgeRewardsSummaryTotalViewController`

scottkicks commented on Kickstarter-iOS/Features/PledgePaymentMethods/Controller/PostCampaignPledgeRewardsSummaryTotalViewController.swift:
```diff
@@ -60,21 +75,67 @@ final class PostCampaignPledgeRewardsSummaryTotalViewController: UIViewControlle
+    self.pledgeOverTimeStackView.isHidden = true
+
+    // TODO: add strings translations [MBL-1860](https://kickstarter.atlassian.net/browse/MBL-1860)
+    self.pledgeOverTimeBadgeLabel.text = "Pledge Over Time"
```
I believe the strings are available on `main` now.

scottkicks commented on Kickstarter-iOS/Features/PledgePaymentMethods/Controller/PostCampaignPledgeRewardsSummaryTotalViewControllerTests.swift:
```diff
@@ -0,0 +1,191 @@
+      }
+  }
+
+  func testView_WithoutReward() {
```
nit: can we update this name to make it clear that it is running with the noShippingAtCheckout feature flag enabled?

scottkicks reviewed (changes requested):
Nice! I made some `nit` suggestions and I think the latest string translations were merged into `main` while I was out so we can update the hardcoded ones now.

jovaniks commented on Kickstarter-iOS/Features/PledgeView/Controllers/PledgeSummaryViewController.swift:
```diff
@@ -101,6 +101,7 @@ final class PledgeSummaryViewController: UIViewController {
 
   internal func configure(with data: PledgeSummaryViewData) {
     self.viewModel.inputs.configure(with: data)
+    self.viewModel.inputs.configureWith(pledgeOverTimeData: nil)
```
Setting `pledgeOverTimeData` to `nil` is required by the `PledgeSummaryViewModel` to ensure proper configuration. Since this ViewController is not related to Pledge Over Time, this change ensures the confirmation text label displays the correct content for non-Pledge Over Time scenarios.

amy-at-kickstarter commented on Kickstarter-iOS/Features/PledgePaymentMethods/Controller/NoShippingPledgeRewardsSummaryViewController.swift:
```diff
@@ -169,6 +169,10 @@ final class NoShippingPledgeRewardsSummaryViewController: UIViewController {
   }
 
+  func configureWith(pledgeOverTimeData: PledgePaymentPlansAndSelectionData?) {
+    self.pledgeTotalViewController.configureWith(pledeOverTimeData: pledgeOverTimeData)
```
Nit: typo (`plede` vs `pledge`)

amy-at-kickstarter commented on Kickstarter-iOS/Features/PledgePaymentMethods/Controller/PledgeRewardsSummaryTotalViewController.swift:
```diff
@@ -93,13 +153,20 @@ final class PostCampaignPledgeRewardsSummaryTotalViewController: UIViewControlle
   }
+
+  internal func configureWith(pledeOverTimeData: PledgePaymentPlansAndSelectionData?) {
+    self.viewModel.inputs.configureWith(pledgeOverTimeData: pledeOverTimeData)
```
Nit, same typo

amy-at-kickstarter commented on Library/ViewModels/PLOTPledgeViewModel.swift:
```diff
@@ -58,16 +85,30 @@ public struct PLOTPledgeViewModel: PLOTPledgeViewModelOutputs {
           thresholdAmount: thresholdAmount
         )
       }
-  }
```
I prefer a solution here that doesn't add an extra boolean to `buildPaymentPlanInputs` - since my intent there was that those arugments would always map 1:1 to `AppEnvironment.current.apiService.buildPaymentPlan`.

Here's an alternative implementation which gets around that:

```
    self.buildPaymentPlanInputs = Signal.combineLatest(project, pledgeTotal)
      // Only call the query once
      .take(first: 1)
      .map({ (project: Project, pledgeTotal: Double) -> (String, String) in
        let amountFormatter = NumberFormatter()
        amountFormatter.minimumFractionDigits = 2
        amountFormatter.maximumFractionDigits = 2
        let amount = amountFormatter.string(from: NSNumber(value: pledgeTotal)) ?? ""

        return (project.slug, amount)
      }
    )
    
    let pledgeOverTimeQuery = self.buildPaymentPlanInputs
      .combineLatest(with: pledgeOverTimeUIEnabled)
      //map ((a, b) c) -> (a, b, c)
      .map({($0.0, $0.1, $1)})
      .switchMap { (slug: String, amount: String, pledgeOverTimeUIEnabled: Bool) -> SignalProducer<
        Signal<GraphAPI.BuildPaymentPlanQuery.Data?, ErrorEnvelope>.Event,
        Never
      > in
        // Proceed with the query only if Pledge Over Time (PLOT) is enabled.
        // If PLOT is disabled, return nil to ensure that the `combineLatest` in `pledgeOverTimeConfigData`
        // emits a value, maintaining the Signal pipeline's flow.
        guard pledgeOverTimeUIEnabled else {
          return SignalProducer(value: .value(nil))
        }

        return AppEnvironment.current.apiService.buildPaymentPlan(
          projectSlug: slug,
          pledgeAmount: amount
        )
        // Wrap the response in an optional and convert the SignalProducer events into materialized values
        // to handle success or error scenarios downstream.
        .wrapInOptional()
        .materialize()
      }
```

amy-at-kickstarter commented on Library/ViewModels/PLOTPledgeViewModel.swift:
```diff
@@ -58,16 +85,30 @@ public struct PLOTPledgeViewModel: PLOTPledgeViewModelOutputs {
-  public let pledgeOverTimeConfigData: Signal<PledgePaymentPlansAndSelectionData, Never>
-  public let buildPaymentPlanInputs: Signal<(String, String), Never>
+    // Sending `.pledgeInFull` as default option
+    self.paymentPlanSelectedObserver.send(value: .pledgeInFull)
```
As far as I know, we don't ever send events direct from a view model on `init`. It would be difficult to prove the order of events - since the correct subscribers in the UI may not be observing yet.

Instead, you should call the input method `paymentPlanSelected` in the `NoShippingPledgeViewController`. Adding it after `self.viewModel.inputs.viewDidLoad()` would be fairly clear, I think.

amy-at-kickstarter reviewed (changes requested):
Mostly LGTM! I have two requested changes for `PLOTPledgeViewModel`, everything else is a nit.

amy-at-kickstarter commented on Kickstarter-iOS/Features/PledgePaymentMethods/Views/PledgeOverTimeBadgeView.swift:
```diff
@@ -0,0 +1,119 @@
+  label.textAlignment = .right
+  label.numberOfLines = 0
+  label.adjustsFontForContentSizeCategory = true
+}
```
👍 Love that this is its own file now, much neater

amy-at-kickstarter commented on Kickstarter-iOS/Features/PledgeView/Controllers/NoShippingPledgeViewController.swift:
```diff
@@ -156,6 +156,9 @@ final class NoShippingPledgeViewController: UIViewController,
     self.viewModel.inputs.viewDidLoad()
+
+    // Sending `.pledgeInFull` as default option
+    self.viewModel.inputs.paymentPlanSelected(.pledgeInFull)
```
👍

amy-at-kickstarter reviewed (approved):
LGTM

scottkicks reviewed (approved):
Thanks for making those changes!
